events/renderers.py

```diff
@@ -12,3 +26,198 @@ class JSONLDRenderer(JSONRenderer):
     media_type = 'application/ld+json'
     format = 'json-ld'
     charset = 'utf-8'
+
+
+class DOCXRenderer(renderers.BaseRenderer):
+    media_type = 'application/vnd.openxmlformats-officedocument.wordprocessingml.document'
+    format = 'docx'
+    charset = None
+    render_style = 'binary'
+
+    def get_query_params(self, renderer_context, param_name):
+        return renderer_context['request'].query_params.get(param_name, '')
+
+    def get_any_language(self, dictionary, default='fi', language_codes=None):
+        if not language_codes:
+            # Default order for when a language isn't found.
+            language_codes = list(set([default, 'fi', 'sv', 'en']))
+
+        for language_code in language_codes:
+            content = dictionary.get(language_code)
+            if content:
+                return content
+        return ''
+
+    def get_date_range(self, renderer_context):
+        dates = {'start': None, 'end': None}
+
+        start_date = self.get_query_params(renderer_context, 'start')
+        if start_date:
+            dates['start'] = dateutil.parser.parse(start_date)
+
+        end_date = self.get_query_params(renderer_context, 'end')
+        if end_date:
+            dates['end'] = dateutil.parser.parse(end_date)
+
+        try:
+            days = int(self.get_query_params(renderer_context, 'days'))
+        except ValueError as e:
+            days = None
+        # Days override dates
+        if days:
+            dates['start'] = datetime.datetime.now().date()
+            dates['end'] = dates['start'] + datetime.timedelta(days=days)
+
+        return dates
+
+    def render(self, data, media_type=None, renderer_context=None):
+        language_code = getattr(settings, 'LANGUAGE_CODE', 'fi')
+        document = Document()
+
+        location = self.get_query_params(renderer_context, 'location')
+        # This could be set to something like "All locations"
+        place_name = ''
+        try:
+            place_name = Place.objects.values_list(
+                'name',
+                flat=True,
+            ).get(id=location)
+        except Place.DoesNotExist:
+            raise Http404('Location not found')
```

Comment:
Hm. Is too much work to actually sort the returned events by date (including year) and location?

This would enable the docx to contain
1) headline for the location, or multiple headlines with events in each location listed under each headline
2) under the location, a headline for the year if the list spans two years, and events of each year in the location listed chronologically. This would visually separate the events if spanning the start of a new year.

These are all obviously edge cases, not the main use case, but if it is reasonably easy to implement, it would be nice that the formatting supported listing e.g. two or more locations separately. This way, any kind of event list could be rendered consistently.